mlir/lib/Dialect/Arith/IR/ArithOps.cpp

```diff
@@ -393,8 +393,12 @@ void arith::AddUIExtendedOp::getCanonicalizationPatterns(
 
 OpFoldResult arith::SubIOp::fold(FoldAdaptor adaptor) {
   // subi(x,x) -> 0
-  if (getOperand(0) == getOperand(1))
-    return Builder(getContext()).getZeroAttr(getType());
+  if (getOperand(0) == getOperand(1)) {
+    auto tensorType = dyn_cast<TensorType>(getType());
```

Comment:
Maybe check for `ShapedType` instead? In case we decide to introduce dynamically-sized vectors one day.

Reply:
Actually, we already kinda have them with scalabale vectors. Can you add test for them too?

Reply:
Good point for using ShapedType.

The scalable vectors aren't an issue I believe since you can create a constant. They act as if they are statically shaped from this point of view.